M2/Macaulay2/packages/Permutations/Code/main.m2

```diff
@@ -0,0 +1,320 @@
+)
+
+
+descents = method()
+descents Permutation := List => w -> (
+    for i in 1 ..< #w list if w_i > w_(i+1) then i else continue
+)
+
+
+ascendingRuns = method()
+ascendingRuns Permutation := List => w -> (
+    -- inspired from the SageMath implementation
+    -- https://github.com/sagemath/sage/blob/develop/src/sage/combinat/permutation.py
+    if #w == 1 then allRuns := {toSequence w}
+    else (
+        allRuns = {};
+        currentRun := {w#0};
+        for wi in w_(toList(1 ..< #w)) do (
+            if wi > last currentRun then currentRun = append(currentRun, wi)
+            else (
+                allRuns = append(allRuns, toSequence currentRun);
+                currentRun = {wi};
+            );
+        );
+        allRuns = append(allRuns, toSequence currentRun);
+    );
+    allRuns
+)
+
+
+descendingRuns = method()
+descendingRuns Permutation := List => w -> (
+    -- inspired from the SageMath implementation
+    -- https://github.com/sagemath/sage/blob/develop/src/sage/combinat/permutation.py
+    if #w == 1 then allRuns := {toSequence w}
+    else (
+        allRuns = {};
+        currentRun := {w#0};
+        for wi in w_(toList(1 ..< #w)) do (
+            if wi < last currentRun then currentRun = append(currentRun, wi)
+            else (
+                allRuns = append(allRuns, toSequence currentRun);
+                currentRun = {wi};
+            );
+        );
+        allRuns = append(allRuns, toSequence currentRun);
+    );
+    allRuns
+)
+
+
+exceedances = method(Options => {Weak => false})
+exceedances Permutation := List => opts -> w -> (
+    compare := if opts.Weak then ((i,j) -> i <= j) else ((i,j) -> i < j);
+    for i in 1 .. #w list if compare(i,w_i) then i else continue 
+)
+
+
+saliances = method()
+saliances Permutation := List => w -> (
+    to1Index positions(1 .. #w, i -> all(i+1 .. #w, j -> w_i > w_j))
+)
+
+
+records = method()
+records Permutation := List => w -> (
+    to1Index positions(1 .. #w, i -> all(1 ..< i, j -> w_j < w_i))
+)
+
+
+------------------------------------
+-- Foata's fundamental bijection
+------------------------------------
+-- see https://en.wikipedia.org/wiki/Permutation#Foata's_transition_lemma
+foataBijection = method()
+foataBijection Permutation := Permutation => w -> (
+    permutation splice cycleDecomposition w
+)
+
+
+------------------------------------
+-- Miscellaneous
+------------------------------------
+-- inverse = method()
+inverse Permutation := Permutation => w -> (permutation to1Index inversePermutation to0Index toList w)
+
+
+-- order of a permutation, i.e. smallest integer n such that w^n = identity
+-- the order of a permutation can be expressed as the lcm of its cycle lengths
+ord = method()
+ord Permutation := ZZ => w -> (
+    lcm((cycleDecomposition w) / length)
+)
+
+
+-- see https://en.wikipedia.org/wiki/Parity_of_a_permutation for different ways
+-- to compute the sign or parity of a permutation
+sign = method()
```

Comment:
Sign `sign` is now exported by Core, this can be removed (and the "sign" in the `export` list).